neutron_fwaas/tests/unit/services/firewall/service_drivers/test_driver_api.py

```diff
@@ -196,7 +200,51 @@ def test_update_firewall_policy(self):
                                        self.driver_api,
                                        payload=self.m_payload)
 
-    # Test Firewall Rule
+    def test_add_router_external_port_to_fwg(self):
+        self.plugin.l3_plugin = mock.Mock()
+        with self.subnet() as subnet:
+            r = router.Router(self.ctx, id=uuidutils.generate_uuid(),
+                                project_id=subnet['subnet']['tenant_id'])
+            r.create()
+            with self.port(subnet=subnet, device_id=r.id, tenant_id=None,
+                            device_owner=nl_constants.DEVICE_OWNER_ROUTER_GW) as gwport:
+
+                with self.firewall_policy() as pol:
+                    with mock.patch.object(self.plugin.l3_plugin, 'get_router',
+                                            return_value=dict(tenant_id=subnet['subnet']['tenant_id'])):
+                        with mock.patch.object(self.plugin.driver, 'is_supported_l3_port',
+                                                return_value=True):
```

Comment:
[optional] If the mocks don't depend on each other then you could consider putting them into the same `with`, separated by `,` to have less indented and therefore better readable code.